### PR TITLE
Relocate HashAdder (Step 1)

### PR DESCRIPTION
**Pull request by mrshenli**
Relocate `HashAdder` for a special case `$Unparallel$ -> HashAdder -> $Parallel$ -> HashRemover` where `HashAdder` is `$Unparallel$`'s only child, and `$Unparallel$` is `HashAdder`'s only parent.

Related to #753

### CODE REVIEW

File: java/src/com/ibm/streamsx/topology/generator/spl/Preprocessor.java [hunk 1]
```diff
@@ -68,4 +71,45 @@ public void compositeColocateIdUsage(List<JsonObject> composites) {
...
     }
+
+    @SuppressWarnings("serial")
+    private void relocateHashAdders(){
+        final Set<JsonObject> hashAdders = new HashSet<>();
+        // Firstly, find each hashAdder
+        GraphUtilities.visitOnce(GraphUtilities.findStarts(graph),
+                new HashSet<BVirtualMarker>(),
+                graph,
+                new Consumer<JsonObject>(){
```
Comment (ddebrunner) [inline java/src/com/ibm/streamsx/topology/generator/spl/Preprocessor.java]: We usually use a lambda expression rather than anonymous class.

File: java/src/com/ibm/streamsx/topology/generator/spl/Preprocessor.java [hunk 2]
```diff
@@ -68,4 +71,45 @@ public void compositeColocateIdUsage(List<JsonObject> composites) {
...
+                    public void accept(JsonObject op) {
+                        if(jstring(op, "kind").equals("com.ibm.streamsx.topology.functional.java::HashAdder")){
```
Comment (ddebrunner) [inline java/src/com/ibm/streamsx/topology/generator/spl/Preprocessor.java]: This is insufficient, and one of the reasons the old code was removed. What is actually needed is a property of the operator invocation to mark it as a logical hash adder, and not determine it from its kind.

File: java/src/com/ibm/streamsx/topology/generator/spl/Preprocessor.java [hunk 3]
```diff
@@ -68,4 +71,45 @@ public void compositeColocateIdUsage(List<JsonObject> composites) {
...
+    private void relocateHashAdder(JsonObject hashAdder){
+        // $Unparallel$ -> hashAdder -> $Parallel$ -> hashRemover
+        Set<JsonObject> parents = GraphUtilities.getUpstream(hashAdder, graph);
+        
+        // hashAdder has multiple parents
+        if (parents.size() != 1) return;
+
+        JsonObject parent = parents.iterator().next();
+
+        // hashAdder's parent has multiple children
```
Comment (ddebrunner) [inline java/src/com/ibm/streamsx/topology/generator/spl/Preprocessor.java]: How could a hash adder have multiple downstream children?

File: java/src/com/ibm/streamsx/topology/generator/spl/Preprocessor.java [hunk 4]
```diff
@@ -68,4 +71,45 @@ public void compositeColocateIdUsage(List<JsonObject> composites) {
...
+        if (GraphUtilities.getDownstream(parent, graph).size() != 1)
+            return;
+
+        // Only optimize the case where $Unparallel has only one child, and
+        // $parallel has only one parent
+        if (jstring(parent, "kind").equals(BVirtualMarker.END_PARALLEL.kind())) {
```
Comment (ddebrunner) [inline java/src/com/ibm/streamsx/topology/generator/spl/Preprocessor.java]: There are utility methods in graph operators to get the kind of an operator invocation.

File: test/java/src/com/ibm/streamsx/topology/test/api/ParallelTest.java [hunk 5]
```diff
@@ -78,8 +78,7 @@ public void fanoutEndParallelException() throws Exception {
 	assertTrue(expectedCount.valid());
     }
 
-    @Test
-    public void testAdjacentParallel() throws Exception {
+    public void testAdjacentParallel(Routing routing1, Routing routing2) throws Exception {
```
Comment (ddebrunner) [inline test/java/src/com/ibm/streamsx/topology/test/api/ParallelTest.java]: private method - only used in this test.

File: .gitignore [hunk 6]
```diff
@@ -43,3 +43,9 @@ samples/java/functional/classes/
 samples/java/functional/functionalsamples.jar
 /.project
 /.pydevproject
+
```
Comment (ddebrunner) [inline .gitignore]: Please don't include unrelated changes in a pr. Submit them as their own pr.

File: java/src/com/ibm/streamsx/topology/generator/spl/Preprocessor.java (hunk 3 above)
Comment (mrshenli) [reply java/src/com/ibm/streamsx/topology/generator/spl/Preprocessor.java]: The **parent** of a `HashAdder` can have multiple children.

File: java/src/com/ibm/streamsx/topology/generator/spl/Preprocessor.java (hunk 2 above)
Comment (mrshenli) [reply java/src/com/ibm/streamsx/topology/generator/spl/Preprocessor.java]: Should I add that property to `OpPerperties` and `BOperatorInvocation#_json()` similar to what is done for `KIND`? Any suggestion for the property name? What about `HASHADDER_OP`?

Comment (ddebrunner) [reply java/src/com/ibm/streamsx/topology/generator/spl/Preprocessor.java]: Boolean JSON property "hashAdder" added in OpProperties.

File: java/src/com/ibm/streamsx/topology/generator/spl/Preprocessor.java (hunk 3 above)
Comment (ddebrunner) [reply java/src/com/ibm/streamsx/topology/generator/spl/Preprocessor.java]: I think I'm confused by the description of this PR which has:

> where HashAdder is $Unparallel's only child **and $parallel's only parent.**

So I thought the check was that the $parallel only has a single upstream operator, which is the hashadder, which will always be the case.

Comment (ddebrunner) [reply java/src/com/ibm/streamsx/topology/generator/spl/Preprocessor.java]: Could you add comments as to why you are putting those restrictions in place, it's fine if it's because only the simple case is being tackled at the moment, but the code is much clearer if a reason is given.

File: java/src/com/ibm/streamsx/topology/generator/spl/Preprocessor.java [hunk 7]
```diff
@@ -68,4 +71,45 @@ public void compositeColocateIdUsage(List<JsonObject> composites) {
...
+        if (parents.size() != 1) return;
+
+        JsonObject parent = parents.iterator().next();
+
+        // hashAdder's parent has multiple children
+        if (GraphUtilities.getDownstream(parent, graph).size() != 1)
+            return;
+
+        // Only optimize the case where $Unparallel has only one child, and
+        // $parallel has only one parent
```
Comment (ddebrunner) [inline java/src/com/ibm/streamsx/topology/generator/spl/Preprocessor.java]: See, again "$parallel has only one parent" but that's not what the code is checking for, so is the code correct and the comment wrong, or vice-versa?

File: java/src/com/ibm/streamsx/topology/generator/spl/Preprocessor.java (hunk 3 above)
Comment (mrshenli) [reply java/src/com/ibm/streamsx/topology/generator/spl/Preprocessor.java]: Sorry, I should have updated the PR description.

File: java/src/com/ibm/streamsx/topology/generator/spl/Preprocessor.java (hunk 2 above)
Comment (mrshenli) [reply java/src/com/ibm/streamsx/topology/generator/spl/Preprocessor.java]: Do you mean this [`OpProperties`](https://github.com/IBMStreams/streamsx.topology/blob/master/java/src/com/ibm/streamsx/topology/generator/operator/OpProperties.java)? I didn't see a `"hashAdder"` property.

File: java/src/com/ibm/streamsx/topology/generator/spl/Preprocessor.java (hunk 7 above)
Comment (ddebrunner) [reply java/src/com/ibm/streamsx/topology/generator/spl/Preprocessor.java]: That's probably the comment I was looking for (assuming it is corrected), but it should be before the checks so that a reader understands the checks as they read down.

File: java/src/com/ibm/streamsx/topology/generator/spl/Preprocessor.java [hunk 8]
```diff
@@ -68,4 +71,45 @@ public void compositeColocateIdUsage(List<JsonObject> composites) {
...
+        if (jstring(parent, "kind").equals(BVirtualMarker.END_PARALLEL.kind())) {
+            GraphUtilities.removeOperator(hashAdder, graph);
```
Comment (ddebrunner) [inline java/src/com/ibm/streamsx/topology/generator/spl/Preprocessor.java]: Does `removeOperator` followed by `addOperator` do the right thing. I think previously `removeOperator` was only used to remove operators that were to be discarded, not added back into the graph. I can't see if the connections for the resultant operator are correct, i.e. does the operator end up with more connections that it should have?

Comment (mrshenli) [reply java/src/com/ibm/streamsx/topology/generator/spl/Preprocessor.java]: this is not addressed yet

File: java/src/com/ibm/streamsx/topology/generator/spl/Preprocessor.java (hunk 2 above)
Comment (mrshenli) [reply java/src/com/ibm/streamsx/topology/generator/spl/Preprocessor.java]: this is not addressed yet

**Comment by mrshenli** [comment]
@ddebrunner 

> Boolean JSON property "hashAdder" added in OpProperties.

I am confused about the above. Do you mean the "hashAdder" property has already been added? I did not find it in [`OpProperties`](https://github.com/IBMStreams/streamsx.topology/blob/master/java/src/com/ibm/streamsx/topology/generator/operator/OpProperties.java). Or do you mean I should add a "hashAdder" property to `BOperatorInvocation#_json()`?

File: java/src/com/ibm/streamsx/topology/generator/spl/Preprocessor.java [hunk 9]
```diff
@@ -68,4 +70,46 @@ public void compositeColocateIdUsage(List<JsonObject> composites) {
...
+        // parent and the HashAdder is $Unparallel$'s only child.
+        // $Unparallel$ -> hashAdder -> $Parallel$ -> hashRemover
+        Set<JsonObject> parents = GraphUtilities.getUpstream(hashAdder, graph);
+        
+        // check if hashAdder has only one parent
+        if (parents.size() != 1) return;
+
+        JsonObject parent = parents.iterator().next();
+
+        // check if $Unparallel$ has only one child
```
Comment (ddebrunner) [inline java/src/com/ibm/streamsx/topology/generator/spl/Preprocessor.java]: FYI - somewhat confusing here because at this point (line 104) it's not guaranteed to be  `$Unparallel$`, since you only check that at line 108. It would actually make more sense to check first it's an `$Unparallel$` before seeing how many downstream operators it has.

File: java/src/com/ibm/streamsx/topology/generator/spl/Preprocessor.java [hunk 10]
```diff
@@ -68,4 +70,46 @@ public void compositeColocateIdUsage(List<JsonObject> composites) {
         for (JsonObject composite : composites)
             pePlacementPreprocess.compositeColocateIdUse(composite);
     }
+
+    @SuppressWarnings("serial")
+    private void relocateHashAdders(){
+        final Set<JsonObject> hashAdders = new HashSet<>();
+        // Firstly, find each hashAdder
+        GraphUtilities.visitOnce(GraphUtilities.findStarts(graph),
```
Comment (ddebrunner) [inline java/src/com/ibm/streamsx/topology/generator/spl/Preprocessor.java]: Is there a reason to collect the list of hash adders first, rather than just relocate them as you discover them?

File: java/src/com/ibm/streamsx/topology/generator/spl/Preprocessor.java [hunk 11]
```diff
@@ -68,4 +70,46 @@ public void compositeColocateIdUsage(List<JsonObject> composites) {
         for (JsonObject composite : composites)
             pePlacementPreprocess.compositeColocateIdUse(composite);
     }
+
+    @SuppressWarnings("serial")
```
Comment (ddebrunner) [inline java/src/com/ibm/streamsx/topology/generator/spl/Preprocessor.java]: Is the suppress warnings needed?

File: java/src/com/ibm/streamsx/topology/generator/spl/Preprocessor.java [hunk 12]
```diff
@@ -68,4 +70,46 @@ public void compositeColocateIdUsage(List<JsonObject> composites) {
...
+
+        JsonObject parent = parents.iterator().next();
+
+        // check if $Unparallel$ has only one child
+        if (GraphUtilities.getDownstream(parent, graph).size() != 1)
+            return;
+
+        if (GraphUtilities.isKind(parent, BVirtualMarker.END_PARALLEL.kind())) {
+            JsonObject hashAdderCopy = GraphUtilities.copyOperatorNewName(
+                    hashAdder, jstring(hashAdder, "name") + "_0");
```
Comment (ddebrunner) [inline java/src/com/ibm/streamsx/topology/generator/spl/Preprocessor.java]: No need to generate a new name, since you are only moving the operator.

File: java/src/com/ibm/streamsx/topology/generator/spl/Preprocessor.java (hunk 10 above)
Comment (ddebrunner) [reply java/src/com/ibm/streamsx/topology/generator/spl/Preprocessor.java]: probably is, due to to not being able to handle graph changes during the iteration.
